opt_adapt/opt.py

```diff
@@ -333,49 +540,46 @@ def minimise(
             if params.disp > 0:
                 pprint(term_msg + "gtol convergence")
             break
+        # For some situation, convergence should be true, but don't satisfy the above condition
+        elif np.abs(J - op.J_progress[-2]) < params.qoi_rtol * op.J_progress[-2]:
+            if fd.norm(dJ) / dJ_init < 1e-3:
+                if params.disp > 0:
+                    pprint(term_msg + "gtol convergence (second situation)")
+                break
 
         # Check for reaching maximum iteration count
         if it == params.maxiter:
             raise fd.ConvergenceError(term_msg + "reaching maxiter")
 
-        # Ramp up the target complexity
-        target = min(target + params.target_inc, params.target_max)
-
-        # Adapt the mesh
-        mesh = adaptor(mesh, target=target, control=u_plus)
-        nc = mesh.num_cells()
-
-        # Check for mesh convergence
-        if adaptor != identity_mesh and np.abs(nc - nc_) < params.element_rtol * nc_:
-            conv = np.array([op.J_progress[i] for i in mesh_conv_it])
-            if (np.abs(J - conv) < params.qoi_rtol * np.abs(conv)).any():
-                pprint(term_msg + "qoi_rtol convergence")
-                break
-            mesh_conv_it.append(it)
-            if params.disp > 1:
-                pprint("NOTE: turning adaptation off due to element_rtol convergence")
-            adaptor = identity_mesh
-            continue
-        else:
-            adaptor = adapt_fn
-        nc_ = nc
-
-        # Check for QoI convergence
-        if it > 1:
+        if it > 1 and mesh_adaptation:
             J_ = op.J_progress[-2]
-            if np.abs(J - J_) < params.qoi_rtol * np.abs(J_):
-                if adaptor != identity_mesh:
-                    conv = np.array([op.J_progress[i] for i in mesh_conv_it])
-                    if (np.abs(J - conv) < params.qoi_rtol * np.abs(conv)).any():
-                        pprint(term_msg + "qoi_rtol convergence")
-                        break
-                mesh_conv_it.append(it)
+            if nc < nc_:
+                mesh_adaptation = False
+                adaptor = identity_mesh
+                if params.disp > 1:
+                    pprint("NOTE: turning adaptation off due to mesh converged")
+            elif np.abs(J - J_) < params.qoi_rtol * np.abs(J_):
+                mesh_adaptation = False
+                adaptor = identity_mesh
                 if params.disp > 1:
                     pprint("NOTE: turning adaptation off due to qoi_rtol convergence")
+            elif np.abs(nc - nc_) < params.element_rtol * nc_:
+                mesh_adaptation = False
                 adaptor = identity_mesh
-                continue
+                if params.disp > 1:
+                    pprint("NOTE: turning adaptation off due to element_rtol convergence")
```

Comment:
Maybe I missed a message somewhere, but can I ask again why you have moved this?